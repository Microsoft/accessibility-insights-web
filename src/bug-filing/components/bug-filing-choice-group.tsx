// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ChoiceGroup, IChoiceGroupOption } from 'office-ui-fabric-react/lib/ChoiceGroup';
import * as React from 'react';

import { UserConfigMessageCreator } from '../../common/message-creators/user-config-message-creator';
import { NamedSFC } from '../../common/react/named-sfc';
import { BugFilingService } from '../types/bug-filing-service';

export type BugFilingChoiceGroupProps = {
    deps: BugFilingChoiceGroupDeps;
    selectedBugFilingService: BugFilingService;
    bugFilingServices: BugFilingService[];
};

export type BugFilingChoiceGroupDeps = {
    userConfigMessageCreator: UserConfigMessageCreator;
};

export const BugFilingChoiceGroup = NamedSFC<BugFilingChoiceGroupProps>('BugFilingChoiceGroup', props => {
    const { userConfigMessageCreator } = props.deps;

    const getOptions: () => IChoiceGroupOption[] = () => {
        return props.bugFilingServices.map(service => {
            return {
                key: service.key,
                text: service.displayName,
            };
        });
    };

    const onChange = (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, option?: IChoiceGroupOption) => {
        userConfigMessageCreator.setBugService(option.key);
    };

    return (
        <ChoiceGroup
            className={'bug-filing-choice-group'}
            onChange={onChange}
            options={getOptions()}
            selectedKey={props.selectedBugFilingService.key}
        />
    );
});
