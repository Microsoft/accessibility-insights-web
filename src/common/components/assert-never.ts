// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// This function helps with exhaustiveness checking in switch statements
// If used in the default case, TS will throw an error unless all cases are handled.
export const assertNever = (x: never): never => {
    throw Error(
        'This error should never be thrown when strict null checks are applied to this project',
    );
};
