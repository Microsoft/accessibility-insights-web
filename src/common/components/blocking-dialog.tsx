// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DefaultButton, PrimaryButton } from 'office-ui-fabric-react/lib/Button';
import { Dialog, DialogFooter, DialogType, IDialogProps } from 'office-ui-fabric-react/lib/Dialog';
import * as React from 'react';

import { NamedSFC } from '../../common/react/named-sfc';

export type BlockingDialogProps = IDialogProps;

export const BlockingDialog = NamedSFC<BlockingDialogProps>('BlockingDialog', props => {
    return (
        // As of writing, NVDA has an issue where it doesn't announce new elements with role=alertdialog,
        // which Office Fabric applies implicitly whenever isBlocking: true is set. As a workaround, our
        // "blocking" dialog sets isBlocking: false (which means the dialog will use role=dialog instead of
        // role=alertdialog, which NVDA announces more reliably), and emulates being "blocking" by making
        // sure we don't set any onDismiss handlers and overriding showCloseButton to false.
        <Dialog
            {...props}
            onDismiss={undefined}
            dialogContentProps={{
                ...props.dialogContentProps,
                showCloseButton: false,
            }}
            modalProps={{
                ...props.modalProps,
                // Be warned that changing this value can change which child element is focused by default
                // (we saw this in the telemetry-permission-dialog); be sure to test this behavior specifically
                // if you ever change this back to true.
                isBlocking: false,
                onDismiss: undefined,
            }}
        />
    );
});
