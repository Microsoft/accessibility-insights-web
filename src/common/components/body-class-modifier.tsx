// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DocumentManipulator } from 'common/document-manipulator';
import * as React from 'react';

export type BodyClassModifierProps = {
    classNames: string[];
    documentManipulator: DocumentManipulator;
};

export class BodyClassModifier extends React.Component<BodyClassModifierProps> {
    public render(): JSX.Element {
        return null;
    }

    private updateClassNamesFrom(prevClassNames: string[]): void {
        const { classNames, documentManipulator } = this.props;

        const addedClassNames = classNames.filter(n => !prevClassNames.includes(n));
        const removedClassNames = prevClassNames.filter(n => !classNames.includes(n));

        // Note: may include class names not managed by us; leave them as-is
        let allBodyClassNames = documentManipulator.bodyClassNames;
        allBodyClassNames = allBodyClassNames.filter(n => !removedClassNames.includes(n));
        allBodyClassNames.push(...addedClassNames);

        documentManipulator.bodyClassNames = allBodyClassNames;
    }

    public componentDidMount(): void {
        this.updateClassNamesFrom([]);
    }

    public componentDidUpdate(prevProps: BodyClassModifierProps): void {
        this.updateClassNamesFrom(prevProps.classNames);
    }
}
