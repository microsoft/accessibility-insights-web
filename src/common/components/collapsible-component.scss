// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

@import '../../common/styles/common.scss';

span.collapsible-title {
    padding-left: 0.5vw;
}

.collapsible {
    padding: 0;
    cursor: pointer;
    width: 100%;
    border: none;
    outline: none;
    line-height: 0.25px;
    display: flex;
    justify-content: flex-start;
    align-items: center;

    h2 {
        font-weight: $normal-title-font-weight;
        font-size: 17px;
    }
}

.collapsible-content {
    overflow: hidden;
    padding-left: calc(0.5vw + 13px); // 13px is Chevron width.
}
