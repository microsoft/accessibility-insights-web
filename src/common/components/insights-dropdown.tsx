// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Dropdown, IDropdownOption, IDropdownProps } from '@fluentui/react';
import * as React from 'react';
import { NamedFC } from '../react/named-fc';

export type InsightsDropdownProps = IDropdownProps & {
    selectedKey: string | number | null;
};

export const InsightsDropdown = NamedFC<InsightsDropdownProps>('InsightsDropdown', props => {
    const { options, selectedKey } = props;

    const addSelectedToAriaLabel = (option: IDropdownOption): IDropdownOption => {
        const isSelected = selectedKey === option.key;
        const screenreaderText = option.ariaLabel ?? option.text;
        return { ...option, ariaLabel: isSelected ? `${screenreaderText} selected` : undefined };
    };

    const modifiedProps = {
        ...props,
        options: options.map(addSelectedToAriaLabel),
    };

    return <Dropdown {...modifiedProps} />;
});
