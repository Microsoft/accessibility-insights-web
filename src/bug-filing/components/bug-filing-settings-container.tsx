// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedSFC } from '../../common/react/named-sfc';
import { BugServiceProperties } from '../../common/types/store-data/user-configuration-store';
import { SettingsDeps } from '../../DetailsView/components/settings-panel/settings/settings-props';
import { BugFilingServiceProvider } from '../bug-filing-service-provider';
import { BugFilingService } from '../types/bug-filing-service';
import { BugFilingChoiceGroup, BugFilingChoiceGroupDeps } from './bug-filing-choice-group';

export interface BugFilingSettingsContainerProps {
    deps: BugFilingSettingsContainerDeps;
    selectedBugFilingService: BugFilingService;
    selectedBugFilingServiceData: BugServiceProperties;
}

export type BugFilingSettingsContainerDeps = {
    bugServiceProvider: BugFilingServiceProvider;
} & BugFilingChoiceGroupDeps &
    SettingsDeps;

export const BugFilingSettingsContainer = NamedSFC<BugFilingSettingsContainerProps>('BugFilingSettingsContainer', props => {
    const { deps, selectedBugFilingService, selectedBugFilingServiceData } = props;
    const SettingsForm = selectedBugFilingService.renderSettingsForm;
    const bugFilingServices = deps.bugServiceProvider.allVisible();

    return (
        <>
            <BugFilingChoiceGroup deps={deps} selectedBugFilingService={selectedBugFilingService} bugFilingServices={bugFilingServices} />
            <SettingsForm deps={deps} settings={selectedBugFilingServiceData} />
        </>
    );
});
