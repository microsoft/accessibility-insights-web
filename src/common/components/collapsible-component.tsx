// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ActionButton } from '@fluentui/react';
import { css } from '@fluentui/utilities';
import * as styles from 'common/components/collapsible-component.scss';
import * as React from 'react';

export interface CollapsibleComponentProps {
    header: JSX.Element;
    content: JSX.Element;
    contentClassName?: string;
    containerClassName?: string;
}

interface CollapsibleComponentState {
    showContent: boolean;
}

export class CollapsibleComponent extends React.Component<
    CollapsibleComponentProps,
    CollapsibleComponentState
> {
    private readonly iconNameDown = 'ChevronDown';
    private readonly iconNameUp = 'ChevronRight';

    constructor(props: CollapsibleComponentProps) {
        super(props);
        this.state = { showContent: true };
    }

    private onClick = (): void => {
        this.setState(prevState => ({ showContent: !prevState.showContent }));
    };

    public render(): JSX.Element {
        const showContent = this.state.showContent;
        let iconName = this.iconNameUp;
        let content: JSX.Element | null = null;

        if (showContent) {
            iconName = this.iconNameDown;
            content = (
                <div className={css(this.props.contentClassName, styles.collapsibleContent)}>
                    {this.props.content}
                </div>
            );
        }

        return (
            <div className={css(this.props.containerClassName)}>
                <ActionButton
                    className={styles.collapsible}
                    iconProps={{ iconName: iconName }}
                    onClick={this.onClick}
                    aria-expanded={showContent}
                >
                    {this.props.header}
                </ActionButton>
                {content}
            </div>
        );
    }
}
