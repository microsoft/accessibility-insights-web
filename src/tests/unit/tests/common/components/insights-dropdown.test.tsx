// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { IDropdownOption } from '@fluentui/react';
import { InsightsDropdown, InsightsDropdownProps } from 'common/components/insights-dropdown';
import { shallow } from 'enzyme';
import * as React from 'react';

describe('InsightsDropdown', () => {
    const options: IDropdownOption[] = [
        {
            key: 'key1',
            text: 'Item 1',
        },
        {
            key: 'key2',
            text: 'Item 2',
            ariaLabel: 'Existing aria label',
        },
    ];

    it('renders with no item selected', () => {
        const props: InsightsDropdownProps = {
            options,
            selectedKey: null,
            placeHolder: 'placeholder',
        };
        const wrapper = shallow(<InsightsDropdown {...props} />);
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('adds aria-label for selected item with no label', () => {
        const props: InsightsDropdownProps = {
            options,
            selectedKey: 'key1',
            placeHolder: 'placeholder',
        };
        const wrapper = shallow(<InsightsDropdown {...props} />);
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('modifies aria-label for selected item with a label', () => {
        const props: InsightsDropdownProps = {
            options,
            selectedKey: 'key2',
            placeHolder: 'placeholder',
        };
        const wrapper = shallow(<InsightsDropdown {...props} />);
        expect(wrapper.getElement()).toMatchSnapshot();
    });
});
