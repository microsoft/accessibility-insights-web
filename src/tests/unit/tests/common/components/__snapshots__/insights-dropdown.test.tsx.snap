// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InsightsDropdown adds aria-label for selected item with no label 1`] = `
<Dropdown
  options={
    Array [
      Object {
        "ariaLabel": "Item 1 selected",
        "key": "key1",
        "text": "Item 1",
      },
      Object {
        "ariaLabel": undefined,
        "key": "key2",
        "text": "Item 2",
      },
    ]
  }
  placeHolder="placeholder"
  selectedKey="key1"
/>
`;

exports[`InsightsDropdown modifies aria-label for selected item with a label 1`] = `
<Dropdown
  options={
    Array [
      Object {
        "ariaLabel": undefined,
        "key": "key1",
        "text": "Item 1",
      },
      Object {
        "ariaLabel": "Existing aria label selected",
        "key": "key2",
        "text": "Item 2",
      },
    ]
  }
  placeHolder="placeholder"
  selectedKey="key2"
/>
`;

exports[`InsightsDropdown renders with no item selected 1`] = `
<Dropdown
  options={
    Array [
      Object {
        "ariaLabel": undefined,
        "key": "key1",
        "text": "Item 1",
      },
      Object {
        "ariaLabel": undefined,
        "key": "key2",
        "text": "Item 2",
      },
    ]
  }
  placeHolder="placeholder"
  selectedKey={null}
/>
`;
